matrix/matrix.py

```diff
@@ -0,0 +1,198 @@
+import operator
+from itertools import repeat
+
+
+class Matrix:
+    """An immutable object representing a matrix data-type. This
+    implementation only supports even-side matrices (n * n).
+
+    Attributes:
+        _matrix (tuple): contains the data about the matrix. Each line of
+                         the matrix is represented by a tuple.
+    """
+    __slots__ = '_matrix'
+
+    def __init__(self, *lines):
+        if not self._is_valid_data(*lines):
+            raise ValueError('Data is not valid')
```

Comment:
You can and should raise exception with a more indicative message
Let the validity function raise the exception and set the appropriate message (e.g. `"Element {} is of type {}, expected a tuple"`)

Reply:
Fixed in last commit.